src/backend/src/services/slack.services.ts

```diff
@@ -0,0 +1,289 @@
+        case 'message_deleted':
+          //delete the message using the client_msg_id
+          eventMessage = (event as SlackDeletedMessage).previous_message;
+          try {
+            return AnnouncementService.deleteAnnouncement(eventMessage.client_msg_id, organizationId);
+          } catch (ignored) {
+            return;
+          }
+        case 'message_changed':
+          eventMessage = (event as SlackUpdatedMessage).message;
+          break;
+        default:
+          //other events that do not effect announcements
+          return;
+      }
+    } else {
+      eventMessage = event as SlackMessage;
+    }
+
+    //loop through the blocks of the meta data while accumulating the
+    //text and users notified
+    let messageText = '';
+    let userIdsToNotify: string[] = [];
+
+    //Get the settings of all users in this organization to compare slack ids
+    const users = await UsersService.getAllUsers();
+    const userSettings = await Promise.all(
+      users.map((user) => {
+        return UsersService.getUserSettings(user.userId);
+      })
+    );
+
+    //get the name of the user that sent the message from slack
+    let userName: string = '';
+    try {
+      userName = (await getUserName(eventMessage.user)) ?? '';
+    } catch (ignored) {}
+
+    //if slack could not produce the name of the user, look for their name in prisma
+    if (!userName) {
+      const userIdList = userSettings
+        .filter((userSetting) => userSetting.slackId === eventMessage.user)
+        .map((userSettings) => userSettings.userId);
+      if (userIdList.length !== 0) {
+        const prismaUserName = users.find((user) => user.userId === userIdList[0]);
+        userName = prismaUserName
+          ? `${prismaUserName?.firstName} ${prismaUserName?.lastName}`
+          : 'Unknown User:' + eventMessage.user;
+      } else {
+        userName = 'Unknown_User:' + eventMessage.user;
+      }
+    }
+
+    //pull out the blocks of data from the metadata within the message event
+    const richTextBlocks = eventMessage.blocks?.filter((eventBlock: any) => eventBlock.type === 'rich_text');
+
+    if (richTextBlocks && richTextBlocks.length === 1) {
+      for (const element of richTextBlocks[0].elements[0].elements) {
+        messageText += await slackServices.blockToString(element);
+        userIdsToNotify = userIdsToNotify.concat(
+          await slackServices.blockToMentionedUsers(element, userSettings, event.channel)
+        );
+      }
+    } else {
+      return;
+    }
+
+    //get rid of duplicates within the users to notify
+    userIdsToNotify = [...new Set(userIdsToNotify)];
+
+    //if no users are notified, disregard the message
+    if (userIdsToNotify.length === 0) {
+      return;
+    }
+
+    console.log('processed event');
+
+    if (event.subtype === 'message_changed') {
+      //try to edit the announcement, if no announcement with that id exists create a new announcement
+      try {
```

Comment:
get rid of unnecessary try catch

Reply:
My idea behind this try catch was that if we get a message that was edited, but we can't find it in our database, we should create a new message with its content. I think this is an important case as if someone were to make a message but forget to add @ channel, and then edit it to include the broadcast, without this try catch it would get ignored the first time and wouldn't find it so also ignore the edit, but with the try catch it will create a new announcement. LMK if that needs more explaining

Reply:
I think what you could do is just make the update announcements service just return if the original announcement is not found, then you just checked if the updated announcement is defined and if it is not then create a new one

Reply:
oh i see. i misread this orgiinally i thought the create announcement function was wrapped inside an else statement. Yeah we should probably be checking for the right error then because it could fail for other reasons that we wouldn’t want to create the announcement in those scenarios